resources/galera_configuration.rb

```diff
@@ -58,9 +58,36 @@
   end
 end
 
+action :bootstrap do
+  ruby_block 'bootstrap galera cluster' do
+    block do
+      shell_out!('galera_new_cluster')
+      node.normal['mariadb']['galera']['bootstrapped'] = true # rubocop:disable ChefCorrectness/NodeNormal
+    end
+    notifies :stop, "service[#{platform_service_name}]", :before
+    not_if { galera_cluster_bootstrapped? && bootstrapped_attribute_set? }
+  end
+end
+
+action :join do
+  ruby_block 'join galera cluster' do
```

Comment:
I don't understand what this is doing?

Reply:
Essentially it just sets `node['mariadb']['galera']['bootstrapped']` to `true` if the node has joined a cluster, which is needed in order to be found by other nodes using a Chef search.

We attempt to join the cluster by restarting the `mariadb` service on each Chef run until it successfully joins a cluster. Once it's connected, the attribute is set to allow other cluster nodes to find it and update the `gcomm` address in their config file so if/when they restart they have multiple donor nodes to connect.

Hard to test at the moment, but created and destroyed a pair of instances locally a few times and get consistent results; first node runs the `:create` and :bootstrap` actions to bootstrap the cluster and set the attribute, second node runs the `:create` and `:join` actions to restart the service to join node1.